omnisolver/samplers/van.py

```diff
@@ -0,0 +1,188 @@
+                assert not sample.requires_grad
+                assert not x_hat.requires_grad
+                sample_time += time.time() - sample_start_time
+
+                train_start_time = time.time()
+
+                log_prob = self.net.log_prob(sample)
+                with torch.no_grad():
+                    energy = self.ham.energy(sample)
+                    loss = log_prob + beta * energy
+                assert not energy.requires_grad
+                assert not loss.requires_grad
+                loss_reinforce = torch.mean((loss - loss.mean()) * log_prob)
+                loss_reinforce.backward()
+
+                if self.clip_grad > 0:
+                    # nn.utils.clip_grad_norm_(params, self.clip_grad)
+                    parameters = list(filter(lambda p: p.grad is not None, params))
+                    max_norm = float(self.clip_grad)
+                    norm_type = 2
+                    total_norm = 0
+                    for p in parameters:
+                        param_norm = p.grad.data.norm(norm_type)
+                        total_norm += param_norm.item()**norm_type
+                        total_norm = total_norm**(1 / norm_type)
+                        clip_coef = max_norm / (total_norm + self.epsilon)
+                        for p in parameters:
+                            p.grad.data.mul_(clip_coef)
+
+                optimizer.step()
+
+                train_time += time.time() - train_start_time
+
+                if self.print_step and step % self.print_step == 0:
+                    free_energy_mean = loss.mean() / beta / self.n
+                    free_energy_std = loss.std() / beta / self.n
+                    entropy_mean = -log_prob.mean() / self.n
+                    energy_mean = energy.mean() / self.n
+                    mag = sample.mean(dim=0)
+                    mag_mean = mag.mean()
+                    if step > 0:
+                        sample_time /= self.print_step
+                        train_time /= self.print_step
+                    used_time = time.time() - start_time
+                    self.my_log(
+                        'beta = {:.3g}, # {}, F = {:.8g}, F_std = {:.8g}, S = {:.5g}, E = {:.5g}, M = {:.5g}, sample_time = {:.3f}, train_time = {:.3f}, used_time = {:.3f}'
+                        .format(
+                            beta,
+                            step,
+                            free_energy_mean.item(),
+                            free_energy_std.item(),
+                            entropy_mean.item(),
+                            energy_mean.item(),
+                            mag_mean.item(),
+                            sample_time,
+                            train_time,
+                            used_time,
+                        ))
+                    sample_time = 0
+                    train_time = 0
+
+            with open(self.fname, 'a', newline='\n') as f:
```

Comment:
Since we are already returning stuff as `dimod.SampleSet`, how about to get rid of this file-write? Even more, what do we need free energy here for?

Reply:
Again, lifted verbatim from VAN implementation. We will probably need to remove a lot of stuff from here. I think this is related to the logging issues.